dataverse-webapp/src/test/java/edu/harvard/iq/dataverse/dataset/DatasetFieldsInitializerTest.java

```diff
@@ -128,42 +146,94 @@ public void groupAndUpdateEmptyAndRequiredFlag() {
         DatasetVersion datasetVersion = new DatasetVersion();
         datasetVersion.setDataset(dataset);
-        
+
         MetadataBlock citationBlock = makeCitationMetadataBlock();
         MetadataBlock geospatialBlock = makeGeospatialMetadataBlock();
         MetadataBlock socialScienceBlock = makeSocialScienceMetadataBlock();
-        
+
         DatasetField titleField = DatasetField.createNewEmptyDatasetField(makeTitleFieldType(citationBlock), null);
         fillTitle(titleField, "Some Title");
-        
+
         DatasetField authorField = DatasetField.createNewEmptyDatasetField(makeAuthorFieldType(citationBlock), null);
         fillAuthorField(authorField, 0, "John Doe", "John Aff");
-        
+
         DatasetField boundingBoxField = DatasetField.createNewEmptyDatasetField(makeGeographicBoundingBoxFieldType(geospatialBlock), null);
-        
+
         DatasetField unitOfAnalysisField = DatasetField.createNewEmptyDatasetField(makeUnitOfAnalysisFieldType(socialScienceBlock), null);
-        
+
         List<DatasetField> datasetFields = Lists.newArrayList(titleField, authorField, unitOfAnalysisField, boundingBoxField);
         datasetVersion.setDatasetFields(datasetFields);
-        
-        
+
+
         // when
         Map<MetadataBlock, List<DatasetField>> retMetadataBlocks = datasetFieldsInitializer.groupAndUpdateEmptyAndRequiredFlag(datasetFields);
-        
+
         // then
         assertEquals(3, retMetadataBlocks.size());
         assertThat(retMetadataBlocks.keySet(), contains(citationBlock, geospatialBlock, socialScienceBlock));
-        
+
         assertTrue(citationBlock.isHasRequired());
         assertFalse(citationBlock.isEmpty());
         assertEquals(2, retMetadataBlocks.get(citationBlock).size());
-        
+
         assertFalse(geospatialBlock.isHasRequired());
         assertTrue(geospatialBlock.isEmpty());
         assertEquals(1, retMetadataBlocks.get(geospatialBlock).size());
-        
+
         assertFalse(socialScienceBlock.isHasRequired());
         assertTrue(socialScienceBlock.isEmpty());
         assertEquals(1, retMetadataBlocks.get(socialScienceBlock).size());
     }
+
+    @Test
+    public void updateDatasetFieldIncludeFlag() {
+        //given
+        Dataverse dataverse = new Dataverse();
+        Dataset dataset = new Dataset();
+        dataset.setOwner(dataverse);
+        DatasetVersion datasetVersion = new DatasetVersion();
+        List<DatasetField> datasetFields = prepareTestData(dataset, datasetVersion);
+
+        Mockito.when(dataverseFieldTypeInputLevelService.findByDataverseIdAndDatasetFieldTypeIdList(any(), any()))
+                .thenReturn(prepareHiddenFields(Lists.newArrayList(2L)));
+
+        //when
+        List<DatasetField> updatedDsf = datasetFieldsInitializer.updateDatasetFieldIncludeFlag(datasetFields, dataverse);
+
+        //then
+        Assert.assertEquals(3, updatedDsf.stream().filter(DatasetField::isInclude).count());
```

Comment:
TBH i don't really see where do we set DataField.include property to make assertions against it.
I don't see where these assertions come from.
Can you explain and maybe add some comments to the code ?

Reply:
it is set in the `datasetFieldsInitializer.updateDatasetFieldIncludeFlag` method that we are using. I don't think comment would do anything since method name explains the intent.

Reply:
The problem with DatasetField is that the include flag is @ Transient which makes it harder to figure out which field is kept in the db. But that's the problem with most of out Entities.

Reply:
Your're right, but what i meant was: without deep understanding of how datasetFieldsInitializer.updateDatasetFieldIncludeFlag I have no idea reading test code why we assert includes to one value, and !include to the rest. 
Maybe adding more assertions to check not only counts but actual objects we expect to be included/!included would help with that.

Reply:
> without deep understanding of how datasetFieldsInitializer.updateDatasetFieldIncludeFlag I have no idea reading test code why we assert includes to one value, and !include to the rest.

Exactly, the test is like a proof that method works, so in order to fully understand the test you need to understand the method itself first.
Imo checking files one by one wouldn't help much since you still need to understand why one object was false. In order to know that you need to understand the method.

If for example we assume that there is giant 1k line method and we make test for it (which doesn't make sense but that's just for example), we need to understand what's going on inside, how otherwise would you understand what we are trying to proof?

Reply:
Imagine TDD where a test is more like a specification of how a method should act. In this case we don't really care for implementation as long as test cases are met. 
Here saying that one of 4 objects should not be included tells us something, but adding additional assert to specify which is it should be excluded (!FieldType.isIncluded()) and for that reason i would add one more assert to the test.